Introduce follow through behavior for tooltips

vjeux opened the pull request:
Inside of Nuclide, we have multiple places where we have multiple icons close together that have a tooltip: the left toolbar, the bottom status bar, the debugger icons...

The current behavior is very frustrating as you've got to wait for the delay on every single one of them. But, we have a clear signal that the user wants a tooltip when s/he waits the time to see it and it's likely that moving the mouse over the next item quickly means that s/he wants to see it as well.

This pull request introduces the concept of follow through: if you have seen a tooltip, you're going to instantly see tooltip on the next element you mouse over within a short timer (300ms right now).

Test Plan:
Video before:
![](http://g.recordit.co/7PCg0hjohP.gif)

Video after:
![](http://g.recordit.co/9OnZCvy9oI.gif)

Released under CC0

vjeux committed:
Introduce follow through behavior for tooltips

Inside of Nuclide, we have multiple places where we have multiple icons close together that have a tooltip: the left toolbar, the bottom status bar, the debugger icons...

The current behavior is very frustrating as you've got to wait for the delay on every single one of them. But, we have a clear signal that the user wants a tooltip when s/he waits the time to see it and it's likely that moving the mouse over the next item quickly means that s/he wants to see it as well.

This pull request introduces the concept of follow through: if you have seen a tooltip, you're going to instantly see tooltip on the next element you mouse over within a short timer (300ms right now).

Test Plan:
Video before:
![](http://g.recordit.co/7PCg0hjohP.gif)

Video after:
![](http://g.recordit.co/9OnZCvy9oI.gif)

Released under CC0

vjeux commented:
cc @nathansobo

nathansobo commented:
Awesome idea. Are you cool writing some tests for this?

vjeux commented:
Absolutely, let me work on tests. I wasn't sure if you'd be interested in it so didn't want to spend the time upfront. Working on it :)

vjeux commented:
I was checking Sublime to see how it handled this and it seems to implement the same concept:

![](http://g.recordit.co/djrSxSl1Bc.gif)

nathansobo commented:
Yeah it's going to be great. Sorry for this crazy jQuery derived code. It's pretty bad.

vjeux commented on spec/tooltip-manager-spec.coffee:
```diff
@@ -149,6 +160,29 @@ describe "TooltipManager", ->
         manager.add element, title: "Title"
         hover element, ->
-          expect(document.body.querySelector(".tooltip")).toBeDefined()
+          expect(document.body.querySelector(".tooltip")).not.toBeNull()
```
I think it's better to write `.not.toBeNull` because `.toBeDefined` is `.not.toBeUndefined` which is also true if it's `null`.

vjeux commented:
Added a test that didn't pass before and now passes. Let me know if you want me to write more tests :)

Ran it with
```
atom --test spec/tooltip-manager-spec.coffee
```

Commit:
Relocate/rephrase tooltip follow-through test

Also make sure it doesn't leave tooltips on the DOM, which causes
subsequent tests to fail

nathansobo commented:
The test looks great. I relocated it to fit a bit better into the existing test structure and rephrased the description a bit to unpack what "follow-through" means to a newcomer to the code. Once this goes green we can :ship:. Nice addition! Thanks so much.

vjeux commented:
The test failure on appveyor seems spurious

Commit:
:arrow_up: about

nathansobo commented:
Thanks again!
